src/pip/_internal/self_outdated_check.py

```diff
@@ -96,6 +100,102 @@ def was_installed_by_pip(pkg: str) -> bool:
     return dist is not None and "pip" == dist.installer
 
 
+def get_py_executable() -> str:
+    """Get path to launch a Python executable.
+
+    First test if python/python3/pythonX.Y on PATH matches the current
+    interpreter, and use that if possible. Then try to get the correct
+    pylauncher command to launch a process of the current python
+    version, fallback to sys.executable
+    """
+
+    if not sys.executable:
+        # docs (python 3.10) says that sys.executable can be can be None or an
+        # empty string if this value cannot be determined, although this is
+        # very rare. In this case, there is nothing much we can do
+        return "python3"
+
+    # windows paths are case-insensitive, pathlib takes that into account
+    sys_executable_path = Path(sys.executable)
+
+    major, minor, *_ = sys.version_info
+
+    # first handle common case: test if path to python/python3/pythonX.Y
+    # matches sys.executable
+    for py in ("python", "python3", f"python{major}.{minor}"):
+        which = shutil.which(py)
+        if which is None:
+            continue
+
+        try:
+            # resolve() removes symlinks, normalises paths and makes them
+            # absolute
+            if Path(which).resolve() == sys_executable_path.resolve():
+                return py
+
+        except RuntimeError:
+            # happens when resolve() encounters an infinite loop
+            pass
+
+    # version in the format used by pylauncher
+    pylauncher_version = f"-{major}.{minor}-{64 if sys.maxsize > 2**32 else 32}"
+
+    # checks that pylauncher is usable, also makes sure pylauncher recognises
+    # the current python version and has the correct path of the current
+    # executable.
+    try:
+        proc = subprocess.run(
+            ["py", "--list-paths"],
```

Comment:
Not all versions of the launcher support `--list-paths` - how can we be sure the user doesn't have an old version? I guess it just fails to find anything, so that's sort of fine.

Reply:
Isn't that why it's an a try/except block and if that's the case correctly use the `sys.executable` fallback?

Reply:
FYI it appears to be part of the Windows CPython 3.7 release onward and the documented flag to use is `-0p` not `--list-paths`: https://github.com/python/cpython/blob/3.7/Doc/whatsnew/3.7.rst#windows-only-changes

Reply:
Also, upgrading the launcher is optional, so people could end up using a later Python but an old launcher (one reason being if the launcher is installed globally, needing admin rights).

Reply:
Yeah, it would fallback in this case